### ISSUE TITLE
ntdll error on Hangover dll compile

### ISSUE THREAD

**Issue by parkerlreed** [AndreRH/hangover#53]
Previous thread https://github.com/AndreRH/hangover/issues/38

Hardware: Nintendo Switch running Arch ARM aarch64 with L4T 4.9 kernel

Packages: 

```
[parker@alarm hangover]$ uname -a
Linux alarm 4.9.140+ #1 SMP PREEMPT Thu Jan 2 09:32:08 PST 2020 aarch64 GNU/Linux
[parker@alarm hangover]$ pacman -Q `pacman -Qsq mingw-`
mingw-w64-binutils 2.33.1-1
mingw-w64-crt 7.0.0-1
mingw-w64-gcc 9.2.0-1
mingw-w64-headers 7.0.0-1
mingw-w64-winpthreads 7.0.0-1
```

When ntdll is being built as part of the Hangover code path, I am running into a conflict with a couple of the mingw builtins.

```
../../../dlls/ntdll/reg.c:1038:28: error: conflicting types for ‘NtQueryMultipleValueKey’
 1038 | WINBASEAPI NTSTATUS WINAPI NtQueryMultipleValueKey(HANDLE KeyHandle, PKEY_MULTIPLE_VALUE_INFORMATION ListOfValuesToQuery, ULONG NumberOfItems, PVOID MultipleValueInformation, ULONG Length, PULONG ReturnLength)
      |                            ^~~~~~~~~~~~~~~~~~~~~~~
In file included from ../../../dlls/ntdll/reg.c:25:
/usr/i686-w64-mingw32/include/winternl.h:1088:18: note: previous declaration of ‘NtQueryMultipleValueKey’ was here
 1088 |   NTSTATUS NTAPI NtQueryMultipleValueKey(HANDLE KeyHandle, PKEY_VALUE_ENTRY ValueEntries, ULONG EntryCount, PVOID ValueBuffer, PULONG BufferLength, PULONG RequiredBufferLength);
      |                  ^~~~~~~~~~~~~~~~~~~~~~~
../../../dlls/ntdll/reg.c:1176:28: error: conflicting types for ‘NtSetInformationKey’
 1176 | WINBASEAPI NTSTATUS WINAPI NtSetInformationKey(HANDLE KeyHandle, const int KeyInformationClass, PVOID KeyInformation, ULONG KeyInformationLength)
      |                            ^~~~~~~~~~~~~~~~~~~
In file included from ../../../dlls/ntdll/reg.c:25:
/usr/i686-w64-mingw32/include/winternl.h:1089:18: note: previous declaration of ‘NtSetInformationKey’ was here
 1089 |   NTSTATUS NTAPI NtSetInformationKey(HANDLE KeyHandle, KEY_SET_INFORMATION_CLASS KeySetInformationClass, PVOID KeySetInformation, ULONG KeySetInformationLength);
      |                  ^~~~~~~~~~~~~~~~~~~
make[1]: *** [../../../dlls/ntdll/Makefile:18: reg_g.o] Error 1
make[1]: Leaving directory '/home/parker/build/hangover/build/dlls32/ntdll'
make: *** [Makefile:196: build/dlls32/ntdll/ntdll.dll] Error 2
make: *** Waiting for unfinished jobs....
```

I attempted the fixes from https://github.com/AndreRH/hangover/issues/38#issuecomment-572500066 which does get past that error but fails on final linking.

```
Wine build complete.
make[1]: Leaving directory '/home/parker/build/hangover/build/wine-guest32'
ln -sf ../../../dlls64/ntdll/ntdll.dll build/qemu/x86_64-windows-user/qemu_guest_dll64/
make -C build/dlls64/ntdll ntdll.dll
make[1]: Entering directory '/home/parker/build/hangover/build/dlls64/ntdll'
../../../build/wine-host/tools/winegcc/winegcc --winebuild ../../../build/wine-host/tools/winebuild/winebuild -I../../../build/wine-host/include -I../../../wine/include --wine-objdir ../../../build/wine-guest -shared -b x86_64-w64-mingw32 ../../../dlls/ntdll/ntdll.spec *_g.o -o ntdll.dll -nodefaultlibs -nostdlib -Wl,--image-base,0x9000000
/usr/lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld: ntdll.dll-3xxOKd.spec.o:fake:(.edata+0x434): undefined reference to `NtSetContextThread'
/usr/lib/gcc/x86_64-w64-mingw32/9.2.0/../../../../x86_64-w64-mingw32/bin/ld: ntdll.dll-3xxOKd.spec.o:fake:(.edata+0xfb8): undefined reference to `NtSetContextThread'
collect2: error: ld returned 1 exit status
winegcc: x86_64-w64-mingw32-gcc failed
make[1]: *** [../../../dlls/ntdll/Makefile:12: ntdll.dll] Error 2
make[1]: Leaving directory '/home/parker/build/hangover/build/dlls64/ntdll'
make: *** [Makefile:203: build/dlls64/ntdll/ntdll.dll] Error 2
```

**Comment by parkerlreed** [comment]
Here are the changes I made to get to that point (NtSetContext error which appears to be the same as #46)

* dlls/ntdll/reg.c:1038 
    * Remove WINBASEAPI
    * Rename NtQueryMultipleValueKey to qemu_NtQueryMultipleValueKey
    * Rename PKEY_MULTIPLE_VALUE_INFORMATION to PKEY_VALUE_ENTRY

* dlls/ntdll/reg.c:1176
    * Remove WINBASEAPI
    * Rename NtSetInformationKey to qemu_NtSetInformationKey

* dlls/ntdll/ntdll.spec
    * Change both (2) NtQueryMultipleValueKey references to qemu_NtQueryMultipleValueKey
    * Change both (2) NtSetInformationKey references to qemu_NtSetInformationKey

**Comment by Obikawa** [comment]
I confirm this bug exists on Pinebook Pro (Rockchip RK3399, AArch64) loaded with Debian 10 (Testing).

Here is my log for comparison:

`i686-w64-mingw32-gcc ../../../dlls/ntdll/reg.c -c -o reg_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
../../../dlls/ntdll/reg.c:1038:28: error: conflicting types for ‘NtQueryMultipleValueKey’
 WINBASEAPI NTSTATUS WINAPI NtQueryMultipleValueKey(HANDLE KeyHandle, PKEY_MULTIPLE_VALUE_INFORMATION ListOfValuesToQuery, ULONG NumberOfItems, PVOID MultipleValueInformation, ULONG Length, PULONG ReturnLength)
                            ^~~~~~~~~~~~~~~~~~~~~~~
In file included from ../../../dlls/ntdll/reg.c:25:
/usr/share/mingw-w64/include/winternl.h:1088:18: note: previous declaration of ‘NtQueryMultipleValueKey’ was here
   NTSTATUS NTAPI NtQueryMultipleValueKey(HANDLE KeyHandle, PKEY_VALUE_ENTRY ValueEntries, ULONG EntryCount, PVOID ValueBuffer, PULONG BufferLength, PULONG RequiredBufferLength);
                  ^~~~~~~~~~~~~~~~~~~~~~~`
`../../../dlls/ntdll/reg.c:1176:28: error: conflicting types for ‘NtSetInformationKey’
 WINBASEAPI NTSTATUS WINAPI NtSetInformationKey(HANDLE KeyHandle, const int KeyInformationClass, PVOID KeyInformation, ULONG KeyInformationLength)
                            ^~~~~~~~~~~~~~~~~~~`
`In file included from ../../../dlls/ntdll/reg.c:25:
/usr/share/mingw-w64/include/winternl.h:1089:18: note: previous declaration of ‘NtSetInformationKey’ was here
   NTSTATUS NTAPI NtSetInformationKey(HANDLE KeyHandle, KEY_SET_INFORMATION_CLASS KeySetInformationClass, PVOID KeySetInformation, ULONG KeySetInformationLength);
                  ^~~~~~~~~~~~~~~~~~~`

**Comment by parkerlreed** [comment]
I wonder if this is just because of the newer mingw gcc. I think you are on the same version?

**Comment by psstoyanov** [comment]
On the mingw-gcc version - that's plausible. Not sure which version would be considereed "stable" for hangover.

Just out of curiosity, are you guys compiling multithreaded? I've noticed that it can skip certain errors when spawning new jobs and proceed when it should've stopped earlier in the process.

**Comment by Obikawa** [comment]
> Just out of curiosity, are you guys compiling multithreaded? I've noticed that it can skip certain errors when spawning new jobs and proceed when it should've stopped earlier in the process.

Whatever plain "make" does is what I executed. I am going to replace Debian with Manjaro soon, because they managed to break Firefox by using some illegal insctruction in its compilation. If you would like some diagnostic data from me, don't delay.

Last night I cleaned up my repositories to make sure I have exclusively arm64 software installed and started a rebuild from scratch. So far I had one compilation error involving libxslt or libexslt which magically disappeared with another 'make' command issued without any changes.

**Comment by parkerlreed** [comment]
Ive tried both multi and single. For the later test runs when making changes I've only done single.

**Comment by stefand** [comment]
> Here are the changes I made to get to that point (NtSetContext error which appears to be the same as #46)
> 
> * dlls/ntdll/reg.c:1038
>   
>   * Remove WINBASEAPI
>   * Rename NtQueryMultipleValueKey to qemu_NtQueryMultipleValueKey
>   * Rename PKEY_MULTIPLE_VALUE_INFORMATION to PKEY_VALUE_ENTRY
> *

Which one of those helped? In theory your option 2 completely bypasses whatever definition mingw has. Can you send a pull request?

**Comment by parkerlreed** [comment]
@stefand both of those edits are what helped get past that first error. I could try to send something in but I want to at least attempt to get the project running first.

**Comment by parkerlreed** [comment]
As noted in the other thread, we have life! (Albeit not much is actually running. I've been most successful with Skifree)

https://github.com/AndreRH/hangover/issues/46#issuecomment-573702883

**Comment by AndreRH** [comment]
I made some commits, please test again without modifications on current master

**Comment by psstoyanov** [comment]
With current master (with the only [change](https://github.com/AndreRH/hangover/issues/52#issuecomment-602197457) to avoid the Pixman error from #52 ):
```
* ccf0a299 - (HEAD -> master, origin/master, origin/HEAD) Add GitHub action for Linux build (5 days ago) <André Hentschel>
* f6b4f887 - Dockerfile: Add a Linux x86_64 Dockerfile (5 days ago) <André Hentschel>
* 608fc0c9 - Add GitHub action for Android build Inspired by Trass3r (5 days ago) <André Hentschel>
```

There is no change towards the ntdll error:
```
echo "include ../../../dlls/ntdll/Makefile" >> build/dlls32/ntdll/Makefile
ln -sf ../../../dlls32/ntdll/ntdll.dll build/qemu/x86_64-windows-user/qemu_guest_dll32/
make -C build/dlls32/ntdll ntdll.dll
make[1]: Entering directory '/home/pak0st/testing_ground/hangover/build/dlls32/ntdll'
i686-w64-mingw32-gcc ../../../dlls/ntdll/actctx.c -c -o actctx_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/critsection.c -c -o critsection_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/debugtools.c -c -o debugtools_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/directory.c -c -o directory_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/env.c -c -o env_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/error.c -c -o error_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/exception.c -c -o exception_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/file.c -c -o file_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/heap.c -c -o heap_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/loader.c -c -o loader_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/main.c -c -o main_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/misc.c -c -o misc_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/nt.c -c -o nt_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/om.c -c -o om_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/path.c -c -o path_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/printf.c -c -o printf_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/process.c -c -o process_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
i686-w64-mingw32-gcc ../../../dlls/ntdll/reg.c -c -o reg_g.o -I../../../dlls/ntdll/../include -I../../../dlls/ntdll/../../qemu/include -DQEMU_CURRENT_DLL=DLL_NTDLL -O2 -DGUEST_BIT=32 -DHOST_BIT=64 -DNTDDI_VERSION=0x06020000 -D_WIN32_WINNT=0x0603 -DQEMU_DLL_GUEST
../../../dlls/ntdll/reg.c:1038:28: error: conflicting types for ‘NtQueryMultipleValueKey’
 1038 | WINBASEAPI NTSTATUS WINAPI NtQueryMultipleValueKey(HANDLE KeyHandle, PKEY_MULTIPLE_VALUE_INFORMATION ListOfValuesToQuery, ULONG NumberOfItems, PVOID MultipleValueInformation, ULONG Length, PULONG ReturnLength)
      |                            ^~~~~~~~~~~~~~~~~~~~~~~
In file included from ../../../dlls/ntdll/reg.c:25:
/usr/i686-w64-mingw32/include/winternl.h:1088:18: note: previous declaration of ‘NtQueryMultipleValueKey’ was here
 1088 |   NTSTATUS NTAPI NtQueryMultipleValueKey(HANDLE KeyHandle, PKEY_VALUE_ENTRY ValueEntries, ULONG EntryCount, PVOID ValueBuffer, PULONG BufferLength, PULONG RequiredBufferLength);
      |                  ^~~~~~~~~~~~~~~~~~~~~~~
../../../dlls/ntdll/reg.c:1176:28: error: conflicting types for ‘NtSetInformationKey’
 1176 | WINBASEAPI NTSTATUS WINAPI NtSetInformationKey(HANDLE KeyHandle, const int KeyInformationClass, PVOID KeyInformation, ULONG KeyInformationLength)
      |                            ^~~~~~~~~~~~~~~~~~~
In file included from ../../../dlls/ntdll/reg.c:25:
/usr/i686-w64-mingw32/include/winternl.h:1089:18: note: previous declaration of ‘NtSetInformationKey’ was here
 1089 |   NTSTATUS NTAPI NtSetInformationKey(HANDLE KeyHandle, KEY_SET_INFORMATION_CLASS KeySetInformationClass, PVOID KeySetInformation, ULONG KeySetInformationLength);
      |                  ^~~~~~~~~~~~~~~~~~~
make[1]: *** [../../../dlls/ntdll/Makefile:18: reg_g.o] Error 1
make[1]: Leaving directory '/home/pak0st/testing_ground/hangover/build/dlls32/ntdll'
make: *** [Makefile:198: build/dlls32/ntdll/ntdll.dll] Error 2
```

**Comment by AndreRH** [comment]
Is that with mingw-w64 6.0?

**Comment by psstoyanov** [comment]
Info if mingw-w64:
```
~ pacman -Qs mingw-w64 
local/mingw-w64-binutils 2.34-1 (mingw-w64-toolchain mingw-w64)
    Cross binutils for the MinGW-w64 cross-compiler
local/mingw-w64-configure 0.1.1-9
    configure wrapper for MinGW (mingw-w64)
local/mingw-w64-crt-git 7.0.0.20200316-1 (mingw-w64-toolchain mingw-w64)
    MinGW-w64 CRT for Windows (git version)
local/mingw-w64-environment 1-2
    Script providing common environment variables and functions for MinGW (mingw-w64)
local/mingw-w64-gcc 9.3.0-1 (mingw-w64-toolchain mingw-w64)
    Cross GCC for the MinGW-w64 cross-compiler
local/mingw-w64-headers-git 7.0.0.20200311-1 (mingw-w64-toolchain mingw-w64)
    MinGW-w64 headers for Windows (git version)
local/mingw-w64-libpng 1.6.37-1
    A collection of routines used to create PNG format graphics (mingw-w64)
local/mingw-w64-meson 1-18
    Meson wrapper for MinGW (mingw-w64)
local/mingw-w64-pixman 0.38.4-8
    The pixel-manipulation library for X and cairo (mingw-w64)
local/mingw-w64-pkg-config 2-4
    Support pkg-config script (does not depend on any particular version of pkg-config).
    Useful for pkg-config powered cross-compiling with mingw-w64, see
    http://tinyurl.com/pkg-config-cross for more info
local/mingw-w64-winpthreads-git 7.0.0.20200316-1 (mingw-w64-toolchain mingw-w64)
    MinGW-w64 winpthreads library (git version)
local/mingw-w64-zlib 1.2.11-1
    A compression/decompression Library (mingw-w64)
```

**Comment by AndreRH** [comment]
Looks like 7.0
You could uncomment those defines in the mingw headers or downgrade to 6.0

**Comment by AndreRH** [comment]
Or fix it like

**Comment by AndreRH** [comment]
Fixed by 8924f6699aceb99b737580ebdd6bedafc7d6922c
